zarr/_storage/sharded_store.py

```diff
@@ -0,0 +1,102 @@
+from functools import reduce
+from itertools import product
+from typing import Any, Iterable, Iterator, Optional, Tuple
+
+from zarr._storage.store import BaseStore, Store
+from zarr.storage import StoreLike, array_meta_key, attrs_key, group_meta_key
+
+
+def _cum_prod(x: Iterable[int]) -> Iterable[int]:
+    prod = 1
+    yield prod
+    for i in x[:-1]:
+        prod *= i
+        yield prod
+
+
+class ShardedStore(Store):
+    """This class should not be used directly,
+    but is added to an Array as a wrapper when needed automatically."""
+
+    def __init__(
+        self, store:
+        StoreLike,
+        shards: Tuple[int, ...],
+        dimension_separator: str,
+        chunk_has_constant_size: bool,
+        fill_value: bytes,
+        value_len: Optional[int],
+    ) -> None:
+        self._store: BaseStore = BaseStore._ensure_store(store)
+        self._shards = shards
+        # This defines C/F-order
+        self._shards_cumprod = tuple(_cum_prod(shards))
+        self._num_chunks_per_shard = reduce(lambda x, y: x*y, shards, 1)
+        self._dimension_separator = dimension_separator
+        # TODO: add jumptable for compressed data
+        assert not chunk_has_constant_size, "Currently only uncompressed data can be used."
+        self._chunk_has_constant_size = chunk_has_constant_size
+        if not chunk_has_constant_size:
+            assert value_len is not None
+            self._fill_chunk = fill_value * value_len
+        else:
+            self._fill_chunk = None
+
+        # TODO: add warnings for ineffective reads/writes:
+        # * warn if partial reads are not available
+        # * optionally warn on unaligned writes if no partial writes are available
+  
+    def __key_to_sharded__(self, key: str) -> Tuple[str, int]:
+        # TODO: allow to be in a group (aka only use last parts for dimensions)
+        subkeys = map(int, key.split(self._dimension_separator))
```

Comment:
```suggestion
        subkeys = map(int, key.split(self._dimension_separator)[-len(self._shards):])
```

or is it more complicated?

Reply:
It might also need to split by `/`, which separates the groups I think. I'll defer this for now, since it needs more testing and can easily be done later.